### PR TITLE
subscribe to determinate-nixd's uds socket and built-paths endpoint when available

### PR DESCRIPTION
**Pull request by colemickens**
WIP, draft, axum-0.7 branch should merge first!

Missing:
- re-connect in case the socket goes away or stream is interrupted
- more testing
- passing/enqueuing built paths to whatever actually uploads
- handling the BuiltPathProtocol or whatever we call it

### CODE REVIEW

File: magic-nix-cache/src/pbh.rs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                tracing::error!("built-paths: failed to http2 handshake");
+                continue;
+            };
+
+            // NOTE(colemickens): for now we just drop the joinhandle and let it keep running
+            let _join_handle = tokio::task::spawn(async move {
+                if let Err(err) = conn.await {
+                    tracing::error!("Connection failed: {:?}", err);
+                }
+            });
```
Comment (cole-h) [inline magic-nix-cache/src/pbh.rs:48]: Do we need to reconstruct the `conn` every loop...? Or is this something we could only do once (provided it succeeds)?

File: magic-nix-cache/src/pbh.rs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+                // TODO(colemickens): error handling:::
+                let maybe_store_paths = event
+                    .outputs
+                    .iter()
+                    .map(|path| {
+                        state
+                            .store
+                            .follow_store_path(path)
+                            .map_err(|_| anyhow!("ahhhhh"))
```
Comment (cole-h) [inline magic-nix-cache/src/pbh.rs:96]: same

File: magic-nix-cache/src/pbh.rs [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+                let Ok(event): core::result::Result<BuiltPathResponseEventV1, _> =
+                    serde_json::from_slice(event_str)
+                else {
+                    tracing::error!(
+                        "failed to decode built-path response as BuiltPathResponseEventV1"
+                    );
+                    continue;
+                };
+
+                // TODO(colemickens): error handling:::
```
Comment (cole-h) [inline magic-nix-cache/src/pbh.rs:88]: .

File: magic-nix-cache/src/pbh.rs (hunk 2 above)
Comment (grahamc) [inline magic-nix-cache/src/pbh.rs:96]: https://www.youtube.com/watch?v=jn3Vv6VYdxw

File: magic-nix-cache/src/pbh.rs (hunk 1 above)
Comment (colemickens) [reply magic-nix-cache/src/pbh.rs:48]: In this case yes, this was to test dnixd getting rebooted.

**Comment by colemickens** [comment]
Renamed, ready for review. Leaving as draft until we get a chance to test this in a GHA run.

File: magic-nix-cache/src/main.rs [hunk 4]
```diff
@@ -375,6 +325,17 @@ async fn main_cli() -> Result<()> {
...
     });
 
+    let dnixd_uds_socket_dir: &Path = Path::new(&DETERMINATE_STATE_DIR);
+    let dnixd_uds_socket_path = dnixd_uds_socket_dir.join(DETERMINATE_NIXD_SOCKET_NAME);
+
+    if dnixd_uds_socket_path.exists() {
+        crate::pbh::subscribe_uds_post_build_hook(dnixd_uds_socket_path, state.clone()).await?;
+    } else {
+        crate::pbh::setup_legacy_post_build_hook(&args.listen, &mut nix_conf).await?;
+    }
```
Comment (grahamc) [inline magic-nix-cache/src/main.rs:335]:
```suggestion
    if dnixd_uds_socket_path.exists() {
        tracing::info!("Subscribing to Determinate Nixd build events.");
        crate::pbh::subscribe_uds_post_build_hook(dnixd_uds_socket_path, state.clone()).await?;
    } else {
        tracing::info!("Patching nix.conf to use a post-build-hook.");
        crate::pbh::setup_legacy_post_build_hook(&args.listen, &mut nix_conf).await?;
    }
```

**Review by cole-h — APPROVED** [review]

**Pull request merged**